### PR TITLE
FIX : first pass at fixing nbagg close issue

### LINKED ISSUE #4281: nbagg: "Close figure" action does not 'close' the figure object
I am not fully sure how the notebook backend is supposed to work (and I also did not found any documentation on it?), but the following did surprise me a bit:

```
%matplotlib notebook

import matplotlib.pyplot as plt
import numpy as np

t = np.linspace(0, 2*np.pi, 1024)
plt.plot(t, np.sin(t), t, -np.sin(t))
```
The above creates an interactive figure in the notebook. 
But when I close it (the red button with cross) and the figure is displayed in the notebook as a static image, a subsequent plotting call does nothing:

```
plt.plot(t, np.sin(t), t, np.cos(t))
```

It is only when I explicitely create a new figure that it is displayed:
```
plt.figure()
plt.plot(t, np.sin(t), t, np.cos(t))
```
or if I explicitly close the previous figures with e.g. `plt.close('all')`.

So to me it seems like the `plt.plot` call is still seeing the original figure as the current figure and so trying to plot it on this original figure, but as this is already displayed as a static image in the previous cell, this does not do anything.

I don't know if this is on purpose, but e.g. with the qt or gtk backend when you close a figure, `plt.plot` will create a new figure automatically, and not trying to plot on the current figure anymore (as this is closed).

---

matplotlib 1.4.3, ipython 3.0, windows 7

### PR DESCRIPTION
**Pull request by tacaswell**
Add callback to destroy figure when the user hits the red x
in the notebook.

This is not the correct solution, but it (might) work for now.

closes #4281 ?

### CODE REVIEW

**Comment by efiring** [comment]
Why do you say this is not the correct solution?

**Comment by tacaswell** [comment]
The registration of the `Gcf.destroy` method needs to only be done if the user is using pyplot. 

On the other hand, this code path may be only accessed through pyplot so it is ok (and this relates back to my, @OceanWolf , and @fariza long running attempt to pull pyplot related code out of the backends).

**Comment by tacaswell** [comment]
Also because it has both name errors and just does not work as-is (I did this on the train and pushed when I hit the station).

**Comment by tacaswell** [comment]
Fixed and force pushed.

**Comment by tacaswell** [comment]
@tritemio Can you see if this fixes your problem?  If you need this to work for a class very soon we can work out a monkey-patch recipe.

**Comment by tritemio** [comment]
@tacaswell, thanks for fixing this. I have problems building matplotlib on windows. I tried to build a conda package from this recipe [here](https://github.com/conda/conda-recipes/tree/master/matplotlib), replacing the initial part of meta.yml with:

```
package:
  name: matplotlib
  version: 1.5fix

source:
    git_url: https://github.com/tacaswell/matplotlib
    git_tag: fix_nbagg_close
```

but it fails with:

```
C:\Users\laser2002j\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Include\xlocale(342) :
warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
c:\users\laser2002j\anaconda\conda-bld\work\src\ft2font.h(7) : fatal error C1083: Cannot open include file: 'stdint.h': No such file or directory
error: command 'C:\\Users\\laser2002j\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\amd64\\cl.exe' failed with exit status 2
```

Building from sources with `python setup.py build` fails becaus it cannot find freetype and libpng (that are installed through conda). I guess I will try under linux for now.

**Comment by tritemio** [comment]
@tacaswell, I tested your patch on linux. It solves this specific problem, i.e. that no new figure is shown after closing a figure with the "x" button.

However, I have other issues with the `notebook` backend on multiple re-executions of the same cell containing a plot. I don't know if these issues are related. If not, I can move these to a new issue:

1. Procedural API: after the second execution of the same cell no figure is displayed (until the next `close()` call).

2. OO API: after each execution, the old fig and ax objects are overwritten, a new figure is created and the RAM is filled with invisible figures. Even `plt.close('all')` does not free any RAM.

**Comment by tacaswell** [comment]
Those should be new issues.  They are related to the issue is the single
ton nature if the display area in the notebook and the way that ipython
captures refs to outputs.

**Comment by tacaswell** [comment]
This needs to be merged soon.

File: lib/matplotlib/backends/backend_nbagg.py [hunk 1]
```diff
@@ -287,6 +296,7 @@ def on_message(self, message):
         message = json.loads(message['content']['data'])
         if message['type'] == 'closing':
             self.on_close()
+            self.manager.canvas.close_event()
```
Comment (efiring) [inline lib/matplotlib/backends/backend_nbagg.py]: Just curious: why is this down here instead of inside the `on_close()` method?

**Comment by pelson** [comment]
I'm not 100% of the behaviour that is being targeted here. Is this fixing a bug, or changing the behaviour? If the latter, we certainly would want a UAT for this.

> This needs to be merged soon.

I'm confident we can do this by next weekend - is that too late?

File: lib/matplotlib/backends/backend_nbagg.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/backends/backend_nbagg.py]: moved

**Comment by tacaswell** [comment]
The problem is that when you hit the red 'x' the figure is closed on the js side, but not on the mpl side so `plt.gca()` still points to an axes which is not being displayed.

**Comment by OceanWolf** [comment]
It looks fine to me now, MEP27 will fix this anyway, but as the main change occurs only in ``new_figure_manager_given_figure`` which MEP27 doesn't even use, and the only other change in``CommSocket`` looks logical, something I will perhaps need for MEP27.

Note I say this having never worked with ``NbAgg``, only ``WebAgg``, but as far as I understand ``NbAgg`` just extends ``WebAgg``...

**Comment by mdboom** [comment]
This makes sense on first read and I definitely see the problem it's trying to address.  But as the last "nbagg" expert I'll defer to @pelson for the final word on this one.

**Comment by tacaswell** [comment]
Coupled with #4841

**Comment by tacaswell** [comment]
There are two remaining subtle things

 - [ ] make sure figures get closed on kernel restart (also does not happen in 1.4.3)
 - [ ] still get a 'no such comm' error from closing figure programtically from python side

and updating the UAT

**Comment by jenshnielsen** [comment]
Great improvements. Do you want to fix the remaining issues on this branch before merging or do that in a new PR?

I noticed that the connection info in `matplotlib.backends.backend_nbagg.connection_info()` seems to be wrong in interactive mode enabled with the magic. 

I.e. with one figure open and shown it reports the figure a pending show:
```
print(matplotlib.backends.backend_nbagg.connection_info())
Figure 1 - Figure 1
Figures pending show: 1
```

As far as I can tell it reports all figures as pending show in interactive mode. This is however not an issue with this branch but already happens on master.

In IPython/Jupyter 4 `%matplotlib notebook` raises a ShimWarning. We should probably import the comm from ipykernel if IPython 4 is installed. 

```
/usr/local/lib/python2.7/site-packages/IPython/kernel/__init__.py:13: ShimWarning: The `IPython.kernel` package has been deprecated. You should import from ipykernel or jupyter_client instead.
  "You should import from ipykernel or jupyter_client instead.", ShimWarning)
```

**Comment by tacaswell** [comment]
I think both of those can be done in a different PR, but I want to update the UAT before we merge this.

I don't think that the connection_info is useful in interactive mode.  It is looking at `_pylab_helpers._activeQue` which is a list of open figures which the 'active' one last.  When a new figure is set active it is bubbled to the end.  It looks like the point of this is so that when the user closes the active figure we can select the most recently active one (should probably implement this as a heap, but).  In `backend_nbagg.Show` in non-interactive mode reaches is and removes the managers from `_activeQue` (which no other backend does) on show and then uses the length of `_activeQue` to determine how many figures are pending.

**Comment by jenshnielsen** [comment]
:+1:

**Comment by tacaswell** [comment]
This seems to have a bad interaction with some styles, but not all, where
it gets stuck in an infinite redraw loop.

**Comment by tacaswell** [comment]
I can't reproduce the issue I was having on this branch, but saw it on a really hacked up branch (basically merged all the outstanding feature PRs) that I have a demo at pygotham off of.

File: lib/matplotlib/backends/backend_nbagg.py [hunk 2]
```diff
@@ -72,9 +72,10 @@ def connection_info():
     for manager in Gcf.get_all_fig_managers():
         fig = manager.canvas.figure
         result.append('{0} - {0}'.format((fig.get_label() or
-                                        "Figure {0}".format(manager.num)),
-                                       manager.web_sockets))
-    result.append('Figures pending show: {0}'.format(len(Gcf._activeQue)))
+                                          "Figure {0}".format(manager.num)),
+                                         manager.web_sockets))
```
Comment (OceanWolf) [inline lib/matplotlib/backends/backend_nbagg.py]: I assume you did this as a PEP8 change, but the second line still lacks a space of indent...

**Comment by OceanWolf** [comment]
UAT?  What does that mean?

Does your (@tacaswell) comment about ``_pylab_helpers._activeQue``, explain the "careful rebuilding of the activeQue" comment in ``_pylab_helpers``?

**Comment by jenshnielsen** [comment]
User acceptance testing, In the notebook modified in the last commit

**Comment by jenshnielsen** [comment]
Looks like it now needs a rebase otherwise good

**Comment by tacaswell** [comment]
This is a conflict in the UAT, hurray!

**Comment by jenshnielsen** [comment]
I made an issue to track the remaining figure close issues above #4940

**Comment by tacaswell** [comment]
Thanks for taking care of the that issue.

**Pull request merged**